tests/test_scheduler.py

```diff
@@ -0,0 +1,345 @@
+          ("http://foo.com/e", 'c'),
+          ("http://foo.com/f", 'c')]
+
+
+class TestMigration(unittest.TestCase):
+
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp()
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+    def _migration(self, tmp_dir):
+        prev_scheduler_handler = SchedulerHandler()
+        prev_scheduler_handler.priority_queue_cls = 'scrapy.pqueues.ScrapyPriorityQueue'
+        prev_scheduler_handler.jobdir = tmp_dir
+
+        prev_scheduler_handler.create_scheduler()
+        for url in _URLS:
+            prev_scheduler_handler.scheduler.enqueue_request(Request(url))
+        prev_scheduler_handler.close_scheduler()
+
+        next_scheduler_handler = SchedulerHandler()
+        next_scheduler_handler.priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+        next_scheduler_handler.jobdir = tmp_dir
+
+        next_scheduler_handler.create_scheduler()
+
+    def test_migration(self):
+        with self.assertRaises(ValueError):
+            self._migration(self.tmpdir)
+
+
+class TestSchedulerWithDownloaderAwareInMemory(BaseSchedulerInMemoryTester,
+                                               unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+
+    def test_logic(self):
+        downloader = self.mock_crawler.engine.downloader
+        for url, slot in _SLOTS:
+            request = Request(url)
+            downloader._set_slot_key(slot, request, None)
+            self.scheduler.enqueue_request(request)
+
+        slots = list()
+        requests = list()
+        while self.scheduler.has_pending_requests():
+            request = self.scheduler.next_request()
+            slot = downloader._get_slot_key(request, None)
+            slots.append(slot)
+            downloader.increment(slot)
+            requests.append(request)
+        self.assertEqual(len(slots), len(_SLOTS))
+
+        for request in requests:
+            slot = downloader._get_slot_key(request, None)
+            self.mock_crawler.engine.downloader.decrement(slot)
+
+        unique_slots = len(set(s for _, s in _SLOTS))
+        for i in range(0, len(_SLOTS), unique_slots):
+            part = slots[i:i + unique_slots]
+            self.assertEqual(len(part), len(set(part)))
+
+
+def _is_slots_unique(base_slots, result_slots):
+    unique_slots = len(set(s for _, s in base_slots))
+    for i in range(0, len(result_slots), unique_slots):
+        part = result_slots[i:i + unique_slots]
+        assert len(part) == len(set(part))
+
+
+class TestSchedulerWithDownloaderAwareOnDisk(BaseSchedulerOnDiskTester,
+                                             unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'
+
+    def test_logic(self):
+        downloader = self.mock_crawler.engine.downloader
+
+        for url, slot in _SLOTS:
+            request = Request(url)
+            downloader._set_slot_key(slot, request, None)
```

Comment:
As _set_slot_key is not a part of Downloader API, I think it could be better to remove it from MockDownloader, and just create `request = Request(url, meta={'download_slot': slot})` here. `download_slot` is a public API, and it looks like MockDownloader supports it.